lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1217,24 +1217,20 @@ struct StoreOpConversion
                              rewriter.create<LLVM::ConstantOp>(
                                  loc, u32Ty, IntegerAttr::get(u32Ty, elemIdx)));
         }
+#ifdef USE_ROCM
+        llWord = bitcast(llWord, valueElemTy);
+#else
         llWord = bitcast(llWord, valArgTy);
+#endif
         std::string constraint =
             (width == 64) ? "l" : ((width == 32) ? "r" : "c");
         asmArgs.emplace_back(llWord, constraint);
       }
+      
 #ifdef USE_ROCM
-      // TODO: need to revisit this when implementing vectorized stores
-      Value llWord = rewriter.create<LLVM::UndefOp>(loc, wordTy);
-      Value elem = valueElems[vecStart];
-      if (elem.getType().getIntOrFloatBitWidth() <= 8)
-        elem = rewriter.create<LLVM::SExtOp>(loc, type::i8Ty(ctx), elem);
-      Type u32Ty = typeConverter->convertType(type::u32Ty(ctx));
-      llWord =
-          insert_element(wordTy, llWord, elem,
-                         rewriter.create<LLVM::ConstantOp>(
-                             loc, u32Ty, IntegerAttr::get(u32Ty, 0)));
-      llWord = bitcast(llWord, valueElemTy);
-      store(llWord, ptrElems[vecStart]);
+      for (size_t wordIdx = 0; wordIdx < nWords; ++wordIdx) {
```

Comment:
maybe move the store call up into the other loop so something like:

      SmallVector<std::pair<Value, std::string>> asmArgs;
      for (size_t wordIdx = 0; wordIdx < nWords; ++wordIdx) {
        // llWord is a width-len composition
        Value llWord = rewriter.create<LLVM::UndefOp>(loc, wordTy);
        // Insert each value element to the composition
        for (size_t elemIdx = 0; elemIdx < wordNElems; ++elemIdx) {
          const size_t elemOffset = vecStart + wordIdx * wordNElems + elemIdx;
          assert(elemOffset < valueElems.size());
          Value elem = valueElems[elemOffset];
          if (elem.getType().isInteger(1))
            elem = rewriter.create<LLVM::SExtOp>(loc, type::i8Ty(ctx), elem);
          elem = bitcast(elem, valueElemTy);

          Type u32Ty = typeConverter->convertType(type::u32Ty(ctx));
          llWord =
              insert_element(wordTy, llWord, elem,
                             rewriter.create<LLVM::ConstantOp>(
                                 loc, u32Ty, IntegerAttr::get(u32Ty, elemIdx)));
        }
#ifdef USE_ROCM
        llWord = bitcast(llWord, valueElemTy);
        store(llWord, ptrElems[vecStart + wordIdx]);
#else

        llWord = bitcast(llWord, valArgTy);
        std::string constraint =
            (width == 64) ? "l" : ((width == 32) ? "r" : "c");
        asmArgs.emplace_back(llWord, constraint);
#endif
      }
#ifndef USE_ROCM

Reply:
Done